File: Python/static_array.py
class StaticArray:
  """
  attributes:
    data_type: data type of the static array
    lenght: lenght of the array, initially zero 
    max: max length of the array
    data: effectivate storage of the data
  
  methods:
    get_value_at(): return the item at a given index
      ex: array[10, 20, 30]
      var = arr.get_value_at(0)
      now var equals 10
    push(): push a new item at the end of the array
    pop(): delete last item of the array
    view_array(): print all the items of the array beginning from 0
    view_complete_array(): same as 'view_array()', but prints even      void indexes of the array, useful for debug
    reverse_array(): reverse a given array and returns a copy of it
  """
  def __init__(self, data_type, max):
    # it will be by default an array of integers
    self.data_type = data_type
    # lenght of the array
    self.length: int = 0
    # max size of the array
    self.max = max
    # store the values
    self.data = [None] * max

  def get_value_at(self, index):
    return self.data[index] # O(1)

  def push(self, item):
    ret_length: int # cointains the updated length of the array to return,
                    # used to not make the code spaghetti like
    if (type(item) != self.data_type):
      ret_length = self.length
    elif (self.length + 1 <= self.max):
      self.data[self.length] = item
      self.length += 1
      ret_length = self.length
    else:
      print('sorry, index out of range')
      ret_length = self.length
    return ret_length # O(1)

  def pop(self):
    last_item: self.data_type = self.data[self.length - 1]
    self.length -= 1 
    # in a dynamic array we would delete the item
    # instead of setting it to none
    self.data[self.length] = None
    return last_item # O(1)

  def view_array(self):
    i: int = 0
    for i in range(self.length):
      print(f'{i}: ', self.data[i])
    return self.length # O(n)

  def view_complete_array(self):
    i: int = 0
    for i in range(self.max):
      print(f'{i}: ', self.data[i])
    return self.length # O(n)

  # delete an item of the array
  def delete(self, item):
    # finding the item index
    item_index: int = 0
    for i in range(self.length):
      if self.data[i] == item:
        #print(f'{arr.data[i]} is at index of {i}')
        item_index = i

    # changing the array
    for i in range(item_index, self.length - 1):
      self.data[i] = self.data[i + 1]

    self.data[self.length - 1] = None
    self.length -= 1

    return self.length # O(n)

  def reverse_array (self, given_array):
    ret_arr = StaticArray(given_array.data_type, given_array.length)

    i: int = 0
    for i in range(given_array.length -1, -1, -1):
      ret_arr.push(given_array.data[i])

    return ret_arr

# creating static array of 5 spaces
arr = StaticArray(int, 5)
arr.push(11)
arr.push(22)
arr.push(33)
arr.push(44)
arr.push(55)

print('before reversing:')
arr.view_array()
arr = arr.reverse_array(arr)
print('after reversing:')
arr.view_array()

File: C++/arrays/static_array.cpp
#include <iostream>

using std::cout;
using std::endl;

template <class dtype, int max>
class static_array
{
  /*
  Attributes:
  - length: length of the array
  - dtype: data type of the array
  - max: max size of the array
  - data: effectivate data of the static array
  Methods:
  - push(): push a new item at the end of the array
  - get_value_at(): return an item at a given index of the array
  - view_array(): view all the elements of the array
  - pop(): delete the last item of the array
  - reverse_array(): reverse the array itself, takes no arguments
  - void_array(): void the array
  */
  public:
    int length = 0;
    dtype data[max];

    // this first version doesn't need the constructor
    // due the template

    // BEGIN *push* 
    int push (dtype item)
    {
      int ret_length = length; // contains the length to return

      if (length + 1 <= max)
      {
        data[length] = item;
        length++;
        ret_length = length;
      }
      else 
      {
        cout << "sorry, index out of range";
      }

      return ret_length; // O(1)
    }
    // END *push*

    // BEGIN *get_value_at*
    dtype get_value_at(int index)
    {
      return data[index]; // O(1)
    }
    // END *get_value_at*

    void view_array()
    {
      int i;
      for (i = 0; i < length; i++)
      {
        cout << i << ": " << data[i] << endl;
      }

      return;
    }

    dtype pop()
    {
      dtype item = data[length - 1];
      length--;

      data[length] = '\0';

      return item;
    }

    void reverse_array()
    {
      int length_arr = length;
      dtype dummy_arr[length_arr];
      int i; 
      // view_array();

      // creating dummy array
      for (i = 0; i < length_arr; i++)
      {
        dummy_arr[i] = data[i];
      }

      void_array();

      // pushing all the the values of the dummy array
      // into the now voided original
      for (i = length_arr -1; i >= 0; i--)
      {
        push(dummy_arr[i]);
      }

      return; // O(n)
    }

    void void_array()
    {
      int length_arr = length;
      int i;

      for (i = 0; i < length_arr; i++)
      {
        pop();
      }

      return;
    }

};

int main()
{
  static_array<int, 5> arr;

  arr.push(11);
  arr.push(22);
  arr.push(33);

  arr.reverse_array();

  arr.view_array();

 return 0;
}